src/network/networkd-link.c

```diff
@@ -2619,16 +2619,32 @@ static int link_set_ipv6_mtu(Link *link) {
 }
 
 static int link_drop_foreign_config(Link *link) {
-        Address *address;
+        Address *address, *net_address;
         Route *route;
         Iterator i;
         int r;
+        bool skip_address;
 
         SET_FOREACH(address, link->addresses_foreign, i) {
                 /* we consider IPv6LL addresses to be managed by the kernel */
                 if (address->family == AF_INET6 && in_addr_is_link_local(AF_INET6, &address->in_addr) == 1)
                         continue;
 
+                if (link->network) {
+                        skip_address = false;
+                        LIST_FOREACH(addresses, net_address, link->network->static_addresses)
+                                if (address_equal(net_address, address)) {
+                                        r = address_add(link, address->family, &address->in_addr, address->prefixlen, NULL);
+                                        if (r < 0)
+                                                return log_link_error_errno(link, r, "Failed to add address: %m");
+                                        skip_address = true;
+                                        break;
+                                }
```

Comment:
hmm, can't we split the searching loop into its own little helper function that we call and then simply call address_add()  or address_remove() based on its result?